app/controllers/course/reference_timelines_controller.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+class Course::ReferenceTimelinesController < Course::ComponentController
+  load_and_authorize_resource :reference_timeline, through: :course
+
+  def create
+    if @reference_timeline.save
+      render partial: 'reference_timeline', locals: {
+        reference_timeline: @reference_timeline,
+        render_times: reference_timeline_params[:reference_times_attributes].present?
+      }
+    else
+      render json: { errors: @reference_timeline.errors }, status: :bad_request
+    end
+  end
+
+  def update
+    if @reference_timeline.update(reference_timeline_update_params)
+      head :ok
+    else
+      render json: { errors: @reference_timeline.errors }, status: :bad_request
+    end
+  end
+
+  def destroy
+    @alternative_timeline_id = destroy_params[:revert_to]
+    revert_course_users_to_alternative_timeline if @alternative_timeline_id.present?
+
+    ActiveRecord::Base.transaction do
+      if @updated_course_users.present?
+        CourseUser.import! @updated_course_users, on_duplicate_key_update: [:reference_timeline_id]
+        @reference_timeline.course_users.reload
+      end
+
+      @reference_timeline.destroy!
+      head :ok
+    end
+  rescue ActiveRecord::InvalidForeignKey # @alternative_timeline_id is invalid
+    head :bad_request
+  rescue StandardError
+    render json: { errors: @reference_timeline.errors }, status: :bad_request
+  end
+
+  private
+
+  def reference_timeline_params
+    params.require(:reference_timeline).permit(:title, :weight, {
+      reference_times_attributes: Course::ReferenceTimesController.base_params
+    })
+  end
+
+  def reference_timeline_update_params
+    params.require(:reference_timeline).permit(:title, :weight)
+  end
+
+  def revert_course_users_to_alternative_timeline
+    @updated_course_users = []
+
+    @reference_timeline.course_users.each do |course_user|
+      course_user.reference_timeline_id = (@alternative_timeline_id == 'default') ? nil : @alternative_timeline_id
```

Comment:
If the given alternative timeline ID is `nil`, it could mean that the query param `revert_to` is not given. I don't want to have reversion to default timeline be automatic. Because the change of timeline is one-way, I'd be better off if the user explicitly specify the alternative timeline with `revert_to`.

If somehow DELETE was invoked by accident, then the validation will prevent this irreversible (and potentially dangerous) change of timeline from happening.

Regarding the _actual stub value_ we should receive from the user to indicate _reversion to default timeline_, of course we could discuss other strings than `'default'`, but I don't think anything along the lines of `null`, `nil`, or `0` to be ideal. These are our internal representation of what a default timeline is denoted by (assigned to `reference_timeline_id` somewhere).
>Anyone could go in and make it such that `reference_timeline_id` is non-nullable, i.e., even for a default timeline, it stores the `1234` ID of the default timeline record. In this case, `nil` could mean _not assigned_.

It'd probably be better if the stub is easily understandable because it is an API that is exposed to the world.

Aaand, as I was saying this, I recall how `reference_timeline_id: null` in a course user's PATCH body means default timeline.